### PR TITLE
catalog: low hanging fruit for optimizing PostDeserializationChanges

### PR DESCRIPTION
**Pull request by rafiss**
### catprivilege: remove MaybeFixUsagePrivForTablesAndDBs

This fix is no longer needed, since 92ad363 added a 21.2 upgrade job
that fixed this bug for all descriptors.

This is useful because it makes it easier in the future to avoid cloning the
protobuf unless we need to modify it.

---

### internal/catkv: remove redundant RunPostDeserializationChanges call

This function already calls descbuilder.FromSerializedValue, which does
the same thing.

A profile showed that this is very expensive for large numbers of
descriptors.

---

image

Epic: None
Release note: None

### CODE REVIEW

**Comment by postamar** [comment]
Note that until recently, functions like `MaybeFixUsagePrivForTablesAndDBs` were doomed to live on indefinitely because restoring arbitrarily old backups had to be supported. Fortunately we closed that loophole recently by not supporting restores of backups more than 1 release ago.

File: pkg/sql/catalog/internal/catkv/catalog_query.go [hunk 1]
```diff
@@ -191,9 +191,6 @@ func build(
 	if expectedType != catalog.Any && b.DescriptorType() != expectedType {
 		return nil, pgerror.Newf(pgcode.WrongObjectType, "descriptor is a %s", b.DescriptorType())
 	}
-	if err := b.RunPostDeserializationChanges(); err != nil {
-		return nil, errors.NewAssertionErrorWithWrappedErrf(err, "error during RunPostDeserializationChanges")
```
Comment (postamar) [inline pkg/sql/catalog/internal/catkv/catalog_query.go:195]: I wonder if we should move this error wrapping to the `RunPostDeserializationChanges ` call in `descbuilder`. I'm less sure about the assertion error but the message wrapping seems useful; however it's not like we encounter these often in the wild either.

**Review by postamar — APPROVED** [review]
Very nice.

**Comment by rafiss** [comment]
i'll remove the others in a separate PR

bors r+

**Comment by postamar** [comment]
> I guess we do know that runPostDeserializationChangesOnAllDescriptors was run in an upgrade job for v22.2.0 (https://github.com/cockroachdb/cockroach/pull/81310), so that must mean every fix added before that is not needed.

Yes indeed! Also IIRC we didn't only do so in that release, but in some earlier ones as well.

**Pull request merged**